src/lightning/fabric/loggers/csv_logs.py

```diff
@@ -220,12 +221,35 @@ def save(self) -> None:
         if not self.metrics:
             return
 
-        last_m = {}
-        for m in self.metrics:
-            last_m.update(m)
-        metrics_keys = list(last_m.keys())
+        current_keys = _get_keys_from_metrics(self.metrics)
+        new_keys = current_keys - set(self.metrics_keys)
+        self.metrics_keys.extend(new_keys)
 
-        with self._fs.open(self.metrics_file_path, "w", newline="") as f:
-            writer = csv.DictWriter(f, fieldnames=metrics_keys)
-            writer.writeheader()
+        file_exists = self._fs.isfile(self.metrics_file_path)
+
+        if new_keys and file_exists:
+            # we need to re-write the file if the keys (header) change
+            self._rewrite_with_new_header(self.metrics_keys)
+
+        with self._fs.open(self.metrics_file_path, mode=("a" if file_exists else "w"), newline="") as file:
```

Comment:
shouldn't this be under an `else:`?

Reply:
Nope, pretty sure not. The approach is to rewrite the existing file contents, then append the current new rows.